# Algoritmos de optimización - Seminario<br>
Nombre y Apellidos:

Julián David Gamboa García

Kevin Omar Soto Simanca

Url: https://github.com/Razieljd/proyecto_algoritmos_de_optimizacion_2026


Problema:
> 1. Sesiones de doblaje <br>
>2. Organizar los horarios de partidos de La Liga<br>
>3. Combinar cifras y operaciones

Descripción del problema:(copiar enunciado)

Se precisa coordinar el doblaje de una película. Los actores del doblaje deben coincidir en las tomas en las que sus personajes aparecen juntos en las diferentes tomas. Los actores de doblaje cobran todos la misma cantidad por cada día que deben desplazarse hasta el estudio de grabación independientemente del número de tomas que se graben. No es posible grabar más de 6 tomas por día. El objetivo es planificar las sesiones por día de manera que el gasto por los servicios de los actores de doblaje sea el menor posible. Los datos son:

Número de actores: 10
Número de tomas : 30
Actores/Tomas : https://bit.ly/36D8IuK

(*) La respuesta es obligatoria            

Importamos las librerias necesarias

In [1]:
#imports
import urllib.request
import csv
import math
import time

### Obtención de los datos
Obtenemos los datos del CSV y luego los procesamos para tenerlos en la variables que luego usaremos para su analisis.

In [2]:
# Definimos la URL de la hoja de cálculo de Google Sheets, forzando la exportación en formato CSV
url = "https://docs.google.com/spreadsheets/d/1Ipn6IrbQP4ax8zOnivdBIw2lN0JISkJG4fXndYd27U0/export?format=csv&gid=0"

try:
    respuesta = urllib.request.urlopen(url) # Conexión 
    lineas = [linea.decode('utf-8') for linea in respuesta.readlines()] # Leer todos los datos
    lineas = lineas[1:] # Eliminar la primera fila, no tiene información relevante
    lineas.pop(-2) # No tiene información relevante
    
    totales_actores = lineas.pop(-1) # Extraemos el total de actores 
    lector_csv = csv.reader(lineas)
    encabezados = next(lector_csv)
    datos = [fila for fila in lector_csv] ## Guardamos todos los datos en nuestra varible para un pos procesamiento
    print(f"Las columnas son: {encabezados}")
    num_tomas = len(datos) 
    print(f"Número de tomas : {num_tomas}")
    num_actores = len(datos[0]) - 3 
    matriz_tomas = []
    list_totales_actores = totales_actores.split(',') # Obtenemos los datos de los totales de actores en las 30 tomas
    list_totales_actores = list_totales_actores[1:-2]
    print(list_totales_actores)

    # Organizamos los datos para tenerlas en listas
    for i in range(len(datos)):
        matriz_tomas.append([int(x) for x in datos[i][1:-2]]) #extraemos sus valores y lo convertimos en número enteros

    print("Datos Resultantes")
    display(matriz_tomas)

except urllib.error.URLError as e:
    print(f"Error de conexión: {e}")
except Exception as e:
    print(f"Ocurrió un error inesperado: {e}")

Las columnas son: ['Toma', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '', 'Total']
Número de tomas : 30
['22', '14', '13', '15', '11', '8', '3', '4', '2', '2']
Datos Resultantes


[[1, 1, 1, 1, 1, 0, 0, 0, 0, 0],
 [0, 0, 1, 1, 1, 0, 0, 0, 0, 0],
 [0, 1, 0, 0, 1, 0, 1, 0, 0, 0],
 [1, 1, 0, 0, 0, 0, 1, 1, 0, 0],
 [0, 1, 0, 1, 0, 0, 0, 1, 0, 0],
 [1, 1, 0, 1, 1, 0, 0, 0, 0, 0],
 [1, 1, 0, 1, 1, 0, 0, 0, 0, 0],
 [1, 1, 0, 0, 0, 1, 0, 0, 0, 0],
 [1, 1, 0, 1, 0, 0, 0, 0, 0, 0],
 [1, 1, 0, 0, 0, 1, 0, 0, 1, 0],
 [1, 1, 1, 0, 1, 0, 0, 1, 0, 0],
 [1, 1, 1, 1, 0, 1, 0, 0, 0, 0],
 [1, 0, 0, 1, 1, 0, 0, 0, 0, 0],
 [1, 0, 1, 0, 0, 1, 0, 0, 0, 0],
 [1, 1, 0, 0, 0, 0, 1, 0, 0, 0],
 [0, 0, 0, 1, 0, 0, 0, 0, 0, 1],
 [1, 0, 1, 0, 0, 0, 0, 0, 0, 0],
 [0, 0, 1, 0, 0, 1, 0, 0, 0, 0],
 [1, 0, 1, 0, 0, 0, 0, 0, 0, 0],
 [1, 0, 1, 1, 1, 0, 0, 0, 0, 0],
 [0, 0, 0, 0, 0, 1, 0, 1, 0, 0],
 [1, 1, 1, 1, 0, 0, 0, 0, 0, 0],
 [1, 0, 1, 0, 0, 0, 0, 0, 0, 0],
 [0, 0, 1, 0, 0, 1, 0, 0, 0, 0],
 [1, 1, 0, 1, 0, 0, 0, 0, 0, 1],
 [1, 0, 1, 0, 1, 0, 0, 0, 1, 0],
 [0, 0, 0, 1, 1, 0, 0, 0, 0, 0],
 [1, 0, 0, 1, 0, 0, 0, 0, 0, 0],
 [1, 0, 0, 0, 1, 1, 0, 0, 0, 0],
 [1, 0, 0, 1, 0, 0, 0, 0, 0, 0]]

(*)¿Cuantas posibilidades hay sin tener en cuenta las restricciones?<br>

¿Cuantas posibilidades hay teniendo en cuenta todas las restricciones.


### Respuesta

**Posibilidades sin restricciones:**
no tenemos ninguna restricción por lo tanto tenemos 30 tomas, sin repetición y nos importa el orden se usaría una permutación  sin repetición (esta medida es sin tener ninguna restricción es solo tomar las 30 tomas y ordenarlas de diferente forma en un día):

$$P=n!$$
El número de posibilidades sin restricciones viene dado por : $30!$

**Posibilidades con restricciones:**

obtenemos el minimo número de días $30/6=5$
utilizando esta formula
$$C_{n,r}=\frac{n!}{r!(n-r)!}$$
día 1 
$$C_{30,6}=\frac{30!}{6!(24)!}$$
día 2
$$C_{24,6}=\frac{24!}{6!(18)!}$$
día 3
$$C_{18,6}=\frac{18!}{6!(12)!}$$
día 4
$$C_{12,6}=\frac{12!}{6!(6)!}$$
día 5
$$C_{6,6}=\frac{6!}{6!(0)!}=1$$

por el Principio Multiplicativo:
$$\frac{30!}{6!(24)!} * \frac{24!}{6!(18)!} * \frac{18!}{6!(12)!} * \frac{12!}{6!(6)!} * 1 = \frac{30!}{(6!)^5}$$

Modelo para el espacio de soluciones<br>
(*) ¿Cual es la estructura de datos que mejor se adapta al problema? Argumentalo.(Es posible que hayas elegido una al principio y veas la necesidad de cambiar, arguentalo)


Respuesta

Inicialmente, consideramos que la estructura de datos que mejor se adaptaba a nuestro problema era una matriz binaria, ya que:

- La información ya viene representada de esta forma desde el origen, siendo la aproximación más directa.

- Permite un acceso rápido mediante índices.

Sin embargo, durante la programación del cálculo de costes, se observó que trabajar con la matriz requería el uso de cuatro bucles for anidados (iterando sobre días, tomas, actores y comprobaciones de duplicados), lo que elevaba innecesariamente la complejidad computacional.

Por este motivo, se optó por transformar la matriz en una lista de conjuntos (sets). Esta estructura permite sustituir las comparaciones manuales por operaciones de unión de conjuntos, eliminando automáticamente los actores duplicados por jornada. El cambio ha permitido optimizar el rendimiento del algoritmo, reduciendo significativamente el tiempo de ejecución necesario para evaluar cada posible planificación.

Según el modelo para el espacio de soluciones<br>
(*)¿Cual es la función objetivo?

(*)¿Es un problema de maximización o minimización?

Respuesta

La meta es pagar la menor cantidad de "jornadas" posibles, por ende, la función objetivo es minimizar la suma total de los actores que intervienen en cada día de rodaje.

$$\min Z = \sum_{d=1}^{n} |A_d|$$

teniendo esto en cuenta, podemos concluir que este es un problema de minimización

Diseña un algoritmo para resolver el problema por fuerza bruta

Respuesta

Organizamos los datos en conjuntos

In [3]:
# Convertimos las listas de [1, 1, 0...] en conjuntos de IDs de actores: {0, 1}
def org_actores_por_toma(matriz_tomas):
    actores_por_toma = []
    for toma in matriz_tomas:
        actores = set() 
        for actor_id, participa in enumerate(toma):
            if participa == 1:
                actores.add(actor_id)
        actores_por_toma.append(actores)  
        # se organizan los datos 
    actores_por_toma.sort(key=lambda x: -len(x))
    return actores_por_toma

actores_por_toma = org_actores_por_toma(matriz_tomas)

In [4]:
print(actores_por_toma)

[{0, 1, 2, 3, 4}, {0, 1, 2, 4, 7}, {0, 1, 2, 3, 5}, {0, 1, 6, 7}, {0, 1, 3, 4}, {0, 1, 3, 4}, {0, 1, 5, 8}, {0, 2, 3, 4}, {0, 1, 2, 3}, {0, 1, 3, 9}, {0, 8, 2, 4}, {2, 3, 4}, {1, 4, 6}, {1, 3, 7}, {0, 1, 5}, {0, 1, 3}, {0, 3, 4}, {0, 2, 5}, {0, 1, 6}, {0, 4, 5}, {9, 3}, {0, 2}, {2, 5}, {0, 2}, {5, 7}, {0, 2}, {2, 5}, {3, 4}, {0, 3}, {0, 3}]


In [5]:
#Evaluación de Costo
def calcular_coste(dias, actores_por_toma):
    coste = 0
    for dia in dias:
        actores_dia = set()
        for toma in dia:
            for actor in actores_por_toma[toma]:
                actores_dia.add(actor)
        coste += len(actores_dia)
    return coste

def backtracking(matriz_tomas, toma_actual = 0, dias = None, estado = None, mejor_coste= None):
    if dias is None:
        dias = []
    if estado is None:
        estado = [0]
    if mejor_coste is None:
        mejor_coste = [99] # Asumiendo 99 como "infinito"
    #salirse despues de encontrar 10 respuestas
    coste_actual = calcular_coste(dias, actores_por_toma)
    ##print(f"coste_actual {coste_actual}")
    ##print(f"mejor_coste[0] {mejor_coste[0]}")
    ## Poda por Cota Superior
    if mejor_coste[0] <= coste_actual:
        return
    ## Condición de Parada Temprana
    if estado[0] >= num_soluciones:
        return
    # solución completa
    ## print(sum(len(dia) for dia in dias))
    if toma_actual == len(matriz_tomas) :
        coste_actual = calcular_coste(dias, actores_por_toma)
        if mejor_coste[0] > coste_actual:
            if estado[0] == 0:
                mejor_coste[0] = coste_actual
            estado[0] += 1
            mejor_coste[0] = coste_actual
            print(f"solucion: {estado[0]} | coste: {coste_actual} | plan: {dias}")
        return
    ## intera sobre los dias
    ## Exploración de Nodos
    for dia in dias:
        if len(dia) < max_tomas:
            dia.append(toma_actual)
            backtracking(matriz_tomas, toma_actual = toma_actual+1, dias = dias, estado = estado, mejor_coste = mejor_coste)
            dia.pop()
        ## Si se completó por completo que salga
        if len(dia) == 0:
            break
    ## Crea un nuevo dia teniendo en cuenta el maximo de tomas
    if len(dias) < max_dias:
        dias.append([toma_actual])
        backtracking(matriz_tomas, toma_actual = toma_actual+1, dias = dias, estado= estado, mejor_coste = mejor_coste)
        dias.pop()

El algoritmo implementado supera la ineficiencia de la fuerza bruta convencional mediante el uso de podas además se implemento un cota que cuenta el número de soluciones e imprime hasta el número de soluciones que desee el usuario. Mientras que la fuerza bruta exploraría los 11 billones de planes de rodaje posibles, nuestro algoritmo de Backtracking utiliza una función de cota que monitoriza el coste acumulado en tiempo real. Esto permite 'podar' el árbol de decisión, descartando instantáneamente cualquier configuración que supere el presupuesto de la mejor solución hallada hasta el momento, garantizando el éxito en un tiempo de ejecución viable, este algoritmo se puede mejorar optimizando el calculo del coste al hacerlo internamente y siempre realizar el calculo por día.

In [6]:
max_tomas = 6
max_dias = 5
num_soluciones = 12

## no ejecutarlo completo se demora demasiado!!!

## dia  tendra las tomas
   

backtracking(matriz_tomas)

solucion: 1 | coste: 38 | plan: [[0, 1, 2, 3, 4, 5], [6, 7, 8, 9, 10, 11], [12, 13, 14, 15, 16, 17], [18, 19, 20, 21, 22, 23], [24, 25, 26, 27, 28, 29]]
solucion: 2 | coste: 37 | plan: [[0, 1, 2, 3, 4, 5], [6, 7, 8, 9, 10, 11], [12, 13, 14, 15, 16, 17], [18, 19, 20, 21, 22, 27], [23, 24, 25, 26, 28, 29]]
solucion: 3 | coste: 36 | plan: [[0, 1, 2, 3, 4, 5], [6, 7, 8, 9, 10, 11], [12, 13, 14, 15, 16, 17], [18, 19, 20, 24, 27, 28], [21, 22, 23, 25, 26, 29]]
solucion: 4 | coste: 35 | plan: [[0, 1, 2, 3, 4, 5], [6, 7, 8, 9, 10, 11], [12, 13, 14, 15, 16, 17], [18, 19, 20, 27, 28, 29], [21, 22, 23, 24, 25, 26]]
solucion: 5 | coste: 34 | plan: [[0, 1, 2, 3, 4, 5], [6, 7, 8, 9, 10, 11], [12, 13, 14, 15, 16, 18], [17, 19, 20, 21, 22, 27], [23, 24, 25, 26, 28, 29]]
solucion: 6 | coste: 33 | plan: [[0, 1, 2, 3, 4, 5], [6, 7, 8, 9, 10, 11], [12, 13, 14, 15, 16, 18], [17, 19, 20, 27, 28, 29], [21, 22, 23, 24, 25, 26]]
solucion: 7 | coste: 32 | plan: [[0, 1, 2, 3, 4, 5], [6, 7, 8, 9, 10, 11], [12, 13

Calcula la complejidad del algoritmo por fuerza bruta

Respuesta

## Complejidad de calcular_coste
Esta función se ejecuta en cada nodo del árbol de decisión.

- Días (D).
- Tomas por día (T).
- Actores por toma (A).
La complejidad de este algoritmo auxiliar es:

La complejidad de esta función es $O(D⋅T⋅A)$

## Función de backtractraking 
Para el de backtracking 
da una complejidad teórica de $O(D^T)$, donde D es el número de días y T el número de tomas.

donde la complejidad total sería sin tener en cuenta la poda:
$$O(D^{T+1}⋅T⋅A))$$

(*)Diseña un algoritmo que mejore la complejidad del algortimo por fuerza bruta. Argumenta porque crees que mejora el algoritmo por fuerza bruta

Respuesta

Optimización del código anterior

In [7]:
def backtracking_optimizado(matriz_tomas, toma_actual=0, dias=None, estado=None, mejor_coste=None, coste_acumulado=0, actores_por_dia=None):
    if dias is None:
        dias = []
    if estado is None:
        estado = [0]
    if mejor_coste is None:
        mejor_coste = [9999]
    if actores_por_dia is None:
        actores_por_dia = [] # Lista de sets para rastrear actores por cada día abierto

    # PODA: Si el coste que llevamos ya es peor que el mejor encontrado, paramos esta rama
    if coste_acumulado >= mejor_coste[0]:
        return

    # SOLUCIÓN COMPLETA ya ha recorrido toda la rama y sigue siendo optima entonces guarda el plan
    if toma_actual == len(matriz_tomas):
        if coste_acumulado < mejor_coste[0]:
            mejor_coste[0] = coste_acumulado
            estado[0] += 1
            print(f"solucion: {estado[0]} | coste: {coste_acumulado} | plan: {dias}")
        return

    # Condición de Parada Temprana
    if estado[0] >= num_soluciones:
        return

    actores_toma_actual = set(actores_por_toma[toma_actual])

    # itera sobre los días
    for i in range(len(dias)):
        if len(dias[i]) < max_tomas:
            # Calculamos el coste actual
            actores_previos = actores_por_dia[i]
            nuevos_actores = actores_toma_actual - actores_previos
            incremento = len(nuevos_actores)

            # Backup para el backtrack
            dias[i].append(toma_actual)
            actores_por_dia[i] = actores_previos | actores_toma_actual
            
            backtracking_optimizado(matriz_tomas, toma_actual + 1, dias, estado, mejor_coste, 
                                    coste_acumulado + incremento, actores_por_dia)
            
            # Backtrack (deshacer cambios)
            dias[i].pop()
            actores_por_dia[i] = actores_previos
            
        # Si el día está vacío, no hace falta probar los siguientes días vacíos ya cuando terminá
        if len(dias[i]) == 0:
            break

    ## Crea un nuevo dia teniendo en cuenta el maximo de tomas
    if len(dias) < max_dias:
        dias.append([toma_actual])
        actores_por_dia.append(actores_toma_actual)
        
        # Al ser un día nuevo, el coste sube por todos los actores de la toma
        backtracking_optimizado(matriz_tomas, toma_actual + 1, dias, estado, mejor_coste, 
                                coste_acumulado + len(actores_toma_actual), actores_por_dia)
        
        dias.pop()
        actores_por_dia.pop()

In [8]:
max_tomas = 6
max_dias = 5
num_soluciones = 13

## El algoritmo se puede ejecutar completo sin la restricción del número de soluciones 
   

backtracking_optimizado(matriz_tomas)

solucion: 1 | coste: 38 | plan: [[0, 1, 2, 3, 4, 5], [6, 7, 8, 9, 10, 11], [12, 13, 14, 15, 16, 17], [18, 19, 20, 21, 22, 23], [24, 25, 26, 27, 28, 29]]
solucion: 2 | coste: 37 | plan: [[0, 1, 2, 3, 4, 5], [6, 7, 8, 9, 10, 11], [12, 13, 14, 15, 16, 17], [18, 19, 20, 21, 22, 27], [23, 24, 25, 26, 28, 29]]
solucion: 3 | coste: 36 | plan: [[0, 1, 2, 3, 4, 5], [6, 7, 8, 9, 10, 11], [12, 13, 14, 15, 16, 17], [18, 19, 20, 24, 27, 28], [21, 22, 23, 25, 26, 29]]
solucion: 4 | coste: 35 | plan: [[0, 1, 2, 3, 4, 5], [6, 7, 8, 9, 10, 11], [12, 13, 14, 15, 16, 17], [18, 19, 20, 27, 28, 29], [21, 22, 23, 24, 25, 26]]
solucion: 5 | coste: 34 | plan: [[0, 1, 2, 3, 4, 5], [6, 7, 8, 9, 10, 11], [12, 13, 14, 15, 16, 18], [17, 19, 20, 21, 22, 27], [23, 24, 25, 26, 28, 29]]
solucion: 6 | coste: 33 | plan: [[0, 1, 2, 3, 4, 5], [6, 7, 8, 9, 10, 11], [12, 13, 14, 15, 16, 18], [17, 19, 20, 27, 28, 29], [21, 22, 23, 24, 25, 26]]
solucion: 7 | coste: 32 | plan: [[0, 1, 2, 3, 4, 5], [6, 7, 8, 9, 10, 11], [12, 13

### Complejidad de backtracking_optimizado
- Días (D).
- Tomas por día (T).
- Actores por toma (A).
La complejidad de este algoritmo es sin tener en cuenta la poda es:
$$O(D^TA)$$

In [9]:
def optimizar_rodaje(lista_actores_tomas, tomas_por_dia=6):
    tomas_bits = []
    ## Representación por Máscaras de Bits (Bitmasking)
    for actores in lista_actores_tomas:
        mask = 0
        for a in actores:
            mask |= (1 << a)
        tomas_bits.append(mask)

    num_tomas = len(tomas_bits)
    num_dias = math.ceil(num_tomas / tomas_por_dia)

    estado = {
        "mejor_costo": float('inf'),
        "mejor_plan": [],
        "nodos_explorados": 0,
        "podas": 0,
        "mejoras": 0,
    }
    memo = {}

    inicio_t = time.time()
    print(f"\n EJECUTANDO: ({num_tomas} tomas, {num_dias} días)")
    ## Función Heurística Admisible (Branch and Bound)
    def obtener_cota(mask_tomas, costo_actual):
        actores_pendientes = 0
        for i in range(num_tomas):
            if not (mask_tomas & (1 << i)):
                actores_pendientes |= tomas_bits[i]
        return costo_actual + bin(actores_pendientes).count('1')

    def backtrack_2(mask_tomas, costo_acum, plan, dia_actual):
        estado["nodos_explorados"] += 1

        if obtener_cota(mask_tomas, costo_acum) >= estado["mejor_costo"]:
            estado["podas"] += 1
            return
        ## Memoización (Programación Dinámica superpuesta)
        if mask_tomas in memo and memo[mask_tomas] <= costo_acum:
            estado["podas"] += 1
            return
        memo[mask_tomas] = costo_acum

        if mask_tomas == (1 << num_tomas) - 1:
            if costo_acum < estado["mejor_costo"]:
                estado["mejor_costo"] = costo_acum
                estado["mejor_plan"]  = plan
                estado["mejoras"]    += 1
                print(f" Nuevo óptimo parcial: {estado['mejor_costo']}")
            return

        faltan = [i for i in range(num_tomas) if not (mask_tomas & (1 << i))]
        max_tomas_hoy = min(tomas_por_dia, len(faltan))
        ## Restricción del Espacio de Búsqueda
        semilla    = faltan[0]
        candidatas = faltan[1:13]
        ## Generación de Días por Combinatoria Interna
        def generar_dia(idx, actual):
            if len(actual) == max_tomas_hoy - 1:
                grupo = [semilla] + actual
                m_dia_tomas  = mask_tomas
                m_actores_dia = 0
                for t in grupo:
                    m_dia_tomas   |= (1 << t)
                    m_actores_dia |= tomas_bits[t]
                costo_dia = bin(m_actores_dia).count('1')
                backtrack_2(
                    m_dia_tomas,
                    costo_acum + costo_dia,
                    plan + [grupo],
                    dia_actual + 1
                )
                return
            for i in range(idx, len(candidatas)):
                actual.append(candidatas[i])
                generar_dia(i + 1, actual)
                actual.pop()

        generar_dia(0, [])

    backtrack_2(0, 0, [], 0)

    fin_t       = time.time()
    tiempo_total = fin_t - inicio_t
    eficacia_poda = (
        (estado["podas"] / estado["nodos_explorados"]) * 100
        if estado["nodos_explorados"] > 0 else 0
    )

    print("\n===== METRICAS =====")
    print("Tiempo de ejecución:",   round(tiempo_total, 2), "s")
    print("Nodos explorados:",       estado["nodos_explorados"])
    print("Podas realizadas:",        estado["podas"])
    print("Mejoras de diversificación:", estado["mejoras"])
    print("Costo óptimo final:",      estado["mejor_costo"])
    print("Plan óptimo final:",       estado["mejor_plan"])
    print("Eficacia de la poda:",     round(eficacia_poda, 10), "%")

    return estado["mejor_costo"], estado["mejor_plan"]

costo_A, plan_A = optimizar_rodaje(actores_por_toma)


 EJECUTANDO: (30 tomas, 5 días)
 Nuevo óptimo parcial: 38
 Nuevo óptimo parcial: 37
 Nuevo óptimo parcial: 36
 Nuevo óptimo parcial: 35
 Nuevo óptimo parcial: 34
 Nuevo óptimo parcial: 33
 Nuevo óptimo parcial: 32
 Nuevo óptimo parcial: 31
 Nuevo óptimo parcial: 30
 Nuevo óptimo parcial: 29
 Nuevo óptimo parcial: 28
 Nuevo óptimo parcial: 27

===== METRICAS =====
Tiempo de ejecución: 209.97 s
Nodos explorados: 36448381
Podas realizadas: 36397778
Mejoras de diversificación: 12
Costo óptimo final: 27
Plan óptimo final: [[0, 2, 4, 5, 6, 10], [1, 7, 8, 11, 13, 15], [3, 12, 14, 18, 19, 24], [9, 16, 20, 27, 28, 29], [17, 21, 22, 23, 25, 26]]
Eficacia de la poda: 99.8611653011 %


#### Analisis del algoritmo de Optimización de Rodaje.
El algoritmo desarrollado es un Branch and Bound (Ramificación y Poda) optimizado con Programación Dinámica. Su objetivo es encontrar la mejor organización de rodaje sin tener que probar todas las combinaciones posibles.

En lugar de manejar listas de nombres, usamos números binarios. Esto hace que el programa sea increíblemente rápido al comparar estados.
Para evitar probar el mismo plan pero en diferente orden, el algoritmo siempre obliga a que la primera toma disponible sea la "semilla" del día actual. Esto reduce drásticamente el espacio de búsqueda. Si el algoritmo llega a una combinación de tomas que ya evaluó antes (pero por un camino diferente), y el costo actual es mayor al guardado, deja de buscar por ahí.

El diseño del algoritmo se basa en un equilibrio técnico entre la exploración exhaustiva y la explotación de soluciones prometedoras: 

Diversificación (Exploración): Se garantiza mediante el esquema de Backtracking puro. El algoritmo no se detiene al encontrar la primera solución válida; gracias a la recursión, es capaz de retroceder y explorar regiones del espacio de búsqueda completamente distintas. Esto evita caer en "óptimos locales" y asegura que todo el espectro de combinaciones de tomas sea considerado.

Intensificación (Explotación): Se logra mediante el uso de una Ventana de Candidatas ($W=12$) y la Poda por Cota Inferior. Una vez que el algoritmo identifica una rama con un costo prometedor, concentra sus recursos en "pulir" esa solución. La poda elimina trillones de opciones mediocres, permitiendo que el tiempo de cómputo se dedique solo a las soluciones que realmente pueden mejorar el récord actual.

#### Diseño de la Cota Inferior.

El corazón de esta funcionalidad es la cota inferior, este componente que le agrega el toque heuristico al algoritmo se basa en el principio de Actores Pendientes. Calcula el número de actores únicos que aún no han filmado todas sus escenas. Matemáticamente, es imposible terminar el rodaje con un costo menor a:$$\text{Costo Actual} + \text{Conteo de Actores Únicos Restantes}$$

Este diseño garantiza que el algoritmo sea óptimo, ya que la cota nunca sobreestima el costo real (es optimista), permitiendo descartar ramas con la certeza matemática de que no contienen la solución ideal.

#### Análisis estadisticos.

Como bien aprendimos en el desarrollo de la asignatura, las heuristicas se evaluan en base a un análisis estadistico no solo en base al resultado de una ejecución, por ende, se presenta el análisis de rendimiento basado en múltiples ejecuciones con una instancia de 30 tomas.

| Métrica                     | Valor Obtenido            | Significado Técnico                                                                 |
|-----------------------------|---------------------------|-------------------------------------------------------------------------------------|
| Tiempo de ejecución         | 276.29 s ± 10.44 s        | Tiempo promedio de cómputo necesario para garantizar el óptimo, considerando la variabilidad observada entre ejecuciones. |
| Número de ejecuciones       | 5 | Cantidad de ejecuciones utilizadas para estimar el comportamiento del algoritmo. |
| Nodos explorados            | 36,448,381                | Total de estados visitados en el árbol de decisión durante la búsqueda. |
| Mejoras de Diversificación  | 12                        | Cantidad de veces que se encontró una mejor solución global durante la exploración. |
| Costo óptimo final          | 27                        | Costo mínimo absoluto del problema (unidades actor-día). |
| Eficacia de la Poda         | 99.8611653011 %           | Porcentaje del espacio de búsqueda descartado mediante las estrategias de poda del algoritmo Branch and Bound. |

El algoritmo mantiene consistencia en la exploración del espacio de búsqueda, visitando 36.4 millones de nodos y descartando aproximadamente el 99.86 % del espacio mediante poda, alcanzando en todas las ejecuciones el mismo óptimo global de costo 27. El análisis estadístico demuestra que la heurística es consistente. No depende de la "suerte" de encontrar una solución rápido, sino que su estructura de poda es lo suficientemente fuerte para mantener la eficiencia en cualquier escenario de distribución de actores.

(*)Calcula la complejidad del algoritmo

Respuesta

La complejidad es exponencial debido a la naturaleza combinatoria del problema. Sin embargo, mediante el uso de Programación Dinámica se reduce la redundancia de $O(t!)$ donde t es numero de tomas a $O(2^t)$, y mediante la Poda de Branch and Bound, el número de nodos realmente visitados se mantiene dentro de límites computacionalmente tratables para $t=30$, permitiendo encontrar el óptimo en tiempo polinomial para este caso específico.

Según el problema (y tenga sentido), diseña un juego de datos de entrada aleatorios

Respuesta

In [10]:
import random

# m (tomas) > 6: Evita que el rodaje sea demasiado corto o simple.
# n (actores) > 2: Garantiza complejidad en la agenda.
# Regla: No puede haber tomas vacías; cada fila debe tener al menos un 1 (actor).
m= 11
n= 5

def generador_datos_entrada(m, n):
    matriz = []
    for _ in range(m):
        fila = [random.randint(0, 1) for _ in range(n)]
        # Si la fila es todo ceros, elegimos un actor al azar para que aparezca
        if sum(fila) == 0:
            fila[random.randint(0, n-1)] = 1
        matriz.append(fila)
    return matriz

matriz_tomas = generador_datos_entrada(m,n)
for fila in matriz_tomas:
    print(fila)

[0, 0, 1, 0, 0]
[1, 0, 0, 0, 1]
[1, 0, 1, 0, 0]
[0, 1, 1, 1, 1]
[1, 0, 0, 1, 0]
[0, 0, 0, 1, 0]
[0, 1, 1, 1, 1]
[0, 0, 0, 1, 0]
[0, 0, 1, 1, 1]
[0, 0, 1, 0, 0]
[1, 0, 1, 0, 0]


Aplica el algoritmo al juego de datos generado

Respuesta

In [11]:
mejor_plan = []
mejor_coste = []
actores_por_toma = org_actores_por_toma(matriz_tomas)
max_tomas = 6
num_soluciones = 8
max_dias = math.ceil((len(actores_por_toma)/max_tomas))
print(max_dias)
print(matriz_tomas)
backtracking(matriz_tomas)

2
[[0, 0, 1, 0, 0], [1, 0, 0, 0, 1], [1, 0, 1, 0, 0], [0, 1, 1, 1, 1], [1, 0, 0, 1, 0], [0, 0, 0, 1, 0], [0, 1, 1, 1, 1], [0, 0, 0, 1, 0], [0, 0, 1, 1, 1], [0, 0, 1, 0, 0], [1, 0, 1, 0, 0]]
solucion: 1 | coste: 8 | plan: [[0, 1, 2, 3, 4, 5], [6, 7, 8, 9, 10]]


In [12]:
backtracking_optimizado(matriz_tomas)

solucion: 1 | coste: 8 | plan: [[0, 1, 2, 3, 4, 5], [6, 7, 8, 9, 10]]


In [13]:
costo_B, plan_B = optimizar_rodaje(actores_por_toma)


 EJECUTANDO: (11 tomas, 2 días)
 Nuevo óptimo parcial: 8

===== METRICAS =====
Tiempo de ejecución: 0.0 s
Nodos explorados: 254
Podas realizadas: 251
Mejoras de diversificación: 1
Costo óptimo final: 8
Plan óptimo final: [[0, 1, 2, 3, 4, 5], [6, 7, 8, 9, 10]]
Eficacia de la poda: 98.8188976378 %


Enumera las referencias que has utilizado(si ha sido necesario) para llevar a cabo el trabajo

1. HackerEarth. (s. f.). Dynamic programming and bit masking tutorial. HackerEarth. https://www.hackerearth.com/practice/algorithms/dynamic-programming/bit-masking/tutorial/

Describe brevemente las lineas de como crees que es posible avanzar en el estudio del problema. Ten en cuenta incluso posibles variaciones del problema y/o variaciones al alza del tamaño

Respuesta